wasihttp/types.go

```diff
@@ -0,0 +1,180 @@
+		return types.MethodPut()
+	case http.MethodPatch:
+		return types.MethodPatch()
+	case http.MethodDelete:
+		return types.MethodDelete()
+	case http.MethodConnect:
+		return types.MethodConnect()
+	case http.MethodOptions:
+		return types.MethodOptions()
+	case http.MethodTrace:
+		return types.MethodTrace()
+	default:
+		// TODO: is other method allowed? or should we return GET?
+		// https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/net/http/method.go
+		// https://github.com/WebAssembly/wasi-http/blob/main/wit/types.wit#L340C41-L341C69
+		return types.MethodOther(s)
+	}
+}
+
+// path returns the path with query from the [http.Request] URL.
+//
+// if both path and query are empty, set it to None
+// [types.path]: https://github.com/WebAssembly/wasi-http/blob/main/wit/types.wit#L350
+func path(req *http.Request) cm.Option[string] {
+	path := req.URL.RequestURI()
+	if path == "" {
+		return cm.None[string]()
+	}
+	return cm.Some(path)
+}
+
+// toScheme returns the wasi-http [types.toScheme] from the [http.Request] URL.toScheme.
+// If the scheme is not http or https, it returns a [types.SchemeOther].
+//
+// [types.toScheme]: https://github.com/WebAssembly/wasi-http/blob/main/wit/types.wit#L359-L360
+func toScheme(s string) types.Scheme {
+	switch s {
+	case "http":
+		return types.SchemeHTTP()
+	case "https":
+		return types.SchemeHTTPS()
+	default:
+		// TODO: when should we set the scheme to `cm.None` if `req.URL.Scheme` is empty?
+		return types.SchemeOther(s)
+	}
+}
+
+// toBody writes the io.ReadCloser to the wasi-http [types.toBody].
+//
+// [types.toBody]: https://github.com/WebAssembly/wasi-http/blob/v0.2.0/wit/types.wit#L514-L540
+func toBody(body *io.ReadCloser, wasiBody *types.OutgoingBody) error {
+	if body == nil || *body == nil {
+		return nil
+	}
+	defer (*body).Close()
+
+	stream_ := wasiBody.Write()
+	stream := stream_.OK() // the first call should always return OK
+	defer stream.ResourceDrop()
+	if _, err := io.Copy(&outStreamWriter{stream: stream}, *body); err != nil {
+		return fmt.Errorf("wasihttp: %v", err)
+	}
+	return nil
+}
+
+// fromBody reads the wasi-http [types.IncomingBody] to the io.ReadCloser.
+//
+// [types.IncomingBody]: https://github.com/WebAssembly/wasi-http/blob/v0.2.0/wit/types.wit#L397-L427
+func fromBody(body *types.IncomingBody) io.ReadCloser {
+	stream_ := body.Stream()
+	stream := stream_.OK() // the first call should always return OK
+	return &inputStreamReader{stream: stream, incomingBody: body}
+}
+
+type outStreamWriter struct {
+	stream *streams.OutputStream
+}
+
+func (s *outStreamWriter) Write(p []byte) (n int, err error) {
+	res := s.stream.BlockingWriteAndFlush(cm.ToList(p))
+	if res.IsErr() {
+		return 0, fmt.Errorf("wasihttp: %v", res.Err())
+	}
+	return len(p), nil
+}
+
+type inputStreamReader struct {
+	stream       *streams.InputStream
+	incomingBody *types.IncomingBody
+}
+
+func (r *inputStreamReader) Read(p []byte) (n int, err error) {
+	poll := r.stream.Subscribe()
```

Comment:
Ideally this would be handled by Go stdlib where it could block on all concurrent WASI streams.